### PR TITLE
OLH-1664 Add another MFA page.

### PR DESCRIPTION
**Pull request by saralk**
### What changed

Adds a new route (behind a feature flag), which allows the user to select what kind of backup 2fa method they are going to add.

Currently, once selected, the user gets an error. We will not launch this until the entire flow is ready.

Also includes a small refactor, we used to have a partial template for each password change request type. I have refactored this as the only difference was the translation keys.

Screenshot 2024-04-08 at 08 50 17
Screenshot 2024-04-08 at 08 50 27
Screenshot 2024-04-08 at 08 59 22

### CODE REVIEW

File: src/components/add-mfa-method/add-mfa-methods-controller.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { Request, Response, NextFunction } from "express";
+import { PATH_DATA } from "../../app.constants";
+
+export function addMfaMethodGet(req: Request, res: Response): void {
+  const helpText = `<p>${req.t("pages.addMfaMethod.app.help.text1")}</p><p>${req.t("pages.addMfaMethod.app.help.text2")}</p>`;
```
Comment (andrew-moores) [inline src/components/add-mfa-method/add-mfa-methods-controller.ts:5]: Isn't this usually done in the nunjucks template?

File: src/app.ts [hunk 2]
```diff
@@ -21,6 +21,7 @@ import {
   getSessionExpiry,
   getSessionSecret,
   supportActivityLog,
+  supportMfaUpsell,
```
Comment (andrew-moores) [inline src/app.ts:24]: What is the relevance of "Upsell" in the name?

File: src/components/add-mfa-method/add-mfa-methods-controller.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+export function addMfaMethodPost(
+  req: Request,
+  res: Response,
+  next: NextFunction
+): void {
+  const { addMfaMethod } = req.body;
+
+  if (addMfaMethod === "sms") {
```
Comment (andrew-moores) [inline src/components/add-mfa-method/add-mfa-methods-controller.ts:16]: Can "sms" be a constant?  (Same for "app" below).

File: src/components/add-mfa-method/tests/add-mfa-methods-controller.test.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      redirect: sandbox.fake(),
+    };
+    next = sinon.spy();
+  });
+
+  afterEach(() => {
+    sandbox.restore();
+  });
+
+  it("should return 400 status code when addMfaMethod is 'sms'", () => {
```
Comment (andrew-moores) [inline src/components/add-mfa-method/tests/add-mfa-methods-controller.test.ts:29]: Could the name of the test describe what happens from the users perspective?  e.g. should not allow user to add a backup MFA method of APP when the primary method is SMS.

File: src/locales/en/translation.json [hunk 5]
```diff
@@ -181,6 +181,23 @@
...
+      },
+      "app": {
+        "title": "Authenticator app for smartphone, tablet or computer",
+        "help": {
+          "title": "What is an authenticator app?",
+          "text1": "An authenticator app creates a security code that helps confirm it's you when you sign in.",
+          "text2": "You can use an authenticator app on your smartphone, tablet or desktop computer. Download an authenticator app for your smartphone or tablet from your app store or search online for an authenticator app for your computer."
+        }
+      },
+      "backLink": "I do not want to add another way to get security codes",
```
Comment (andrew-moores) [inline src/locales/en/translation.json:198]: Is the design correct that this doesn't end with a full stop?

Comment (saralk) [reply src/locales/en/translation.json:198]: good point, i will clarify that

File: src/components/add-mfa-method/add-mfa-methods-controller.ts (hunk 3 above)
Comment (saralk) [reply src/components/add-mfa-method/add-mfa-methods-controller.ts]: Done, and refactored a bit to use the constants more widely

File: src/components/add-mfa-method/tests/add-mfa-methods-controller.test.ts (hunk 4 above)
Comment (saralk) [reply src/components/add-mfa-method/tests/add-mfa-methods-controller.test.ts]: agreed and changed

**Review by andrew-moores — APPROVED** [review]

**Pull request merged**